implementations/rust/ockam/ockam_transport_tcp/src/ebpf_portal/raw_socket_processor.rs

```diff
@@ -0,0 +1,156 @@
+use crate::ebpf_portal::pnet_helper::{create_raw_socket, next_tcp_packet};
+use crate::ebpf_portal::{
+    Inlet, InletRegistry, Outlet, OutletRegistry, ParsedRawSocketPacket, RawSocketPacket,
+};
+use log::warn;
+use ockam_core::{async_trait, Processor, Result};
+use ockam_node::Context;
+use ockam_transport_core::TransportError;
+use pnet::packet::Packet;
+use pnet::transport::{TransportReceiver, TransportSender};
+use std::sync::{Arc, Mutex};
+use tracing::info;
+
+/// Processor responsible for receiving all data with OCKAM_TCP_PORTAL_PROTOCOL on the machine
+/// and redirect it to individual portal workers.
+pub struct RawSocketProcessor {
+    socket_read_handle: Option<TransportReceiver>,
+
+    inlet_registry: InletRegistry,
+    outlet_registry: OutletRegistry,
+}
+
+impl RawSocketProcessor {
+    pub(crate) async fn create(
+        ip_proto: u8,
+        inlet_registry: InletRegistry,
+        outlet_registry: OutletRegistry,
+    ) -> Result<(Self, Arc<Mutex<TransportSender>>)> {
+        let (socket_write_handle, socket_read_handle) = create_raw_socket(ip_proto)?;
+
+        let s = Self {
+            socket_read_handle: Some(socket_read_handle),
+            inlet_registry,
+            outlet_registry,
+        };
+
+        Ok((s, Arc::new(Mutex::new(socket_write_handle))))
+    }
+
+    async fn handle_inlet(&self, inlet: Inlet, packet: ParsedRawSocketPacket) -> Result<()> {
+        inlet.sender.send(packet).await.unwrap(); //FIXME
+
+        Ok(())
+    }
+
+    async fn handle_outlet(&self, outlet: Outlet, packet: ParsedRawSocketPacket) -> Result<()> {
+        outlet.sender.send(packet).await.unwrap();
+
+        Ok(())
+    }
+
+    async fn get_new_packet(
+        mut socket_read_handle: TransportReceiver,
+    ) -> Result<(TransportReceiver, ParsedRawSocketPacket)> {
+        // TODO: I wonder how bad it is to use blocking here
```

Comment:
I think you can put a timeout  (SO_RCVTIMEO and SO_SNDTIMEO) inside the socket itself rather than manually implementing it. The blocking syscall should return with an error as soon as the timeout is hit.

Reply:
I'm not yet sure what is the correct way to do it. For now, that's mostly copy&pasted code from pnet library. Library only allowed to query packets using an iterator, which was very inconvenient in our context. As the next step I would like to see how reading from a socket/file descriptor is implemented in tokio and reuse that.